compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -257,6 +260,80 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         ty
     }
 
+    /// Whether this expression constitutes a read of value of the type that
+    /// it evaluates to.
+    ///
+    /// This is used to determine if we should consider the block to diverge
+    /// if the expression evaluates to `!`, and if we should insert a `NeverToAny`
+    /// coercion for values of type `!`.
+    ///
+    /// This function generally returns `false` if the expression is a place
+    /// expression and the *parent* expression is the scrutinee of a match or
+    /// the pointee of an `&` addr-of expression, since both of those parent
+    /// expressions take a *place* and not a value.
+    ///
+    /// This function is unfortunately a bit heuristical, though it is certainly
+    /// far sounder than the prior status quo: <https://github.com/rust-lang/rust/issues/117288>.
+    pub(super) fn expr_constitutes_read(&self, expr: &'tcx hir::Expr<'tcx>) -> bool {
+        // We only care about place exprs. Anything else returns an immediate
+        // which would constitute a read. We don't care about distinguishing
+        // "syntactic" place exprs since if the base of a field projection is
+        // not a place then it would've been UB to read from it anyways since
+        // that constitutes a read.
+        if !expr.is_syntactic_place_expr() {
+            return true;
+        }
+
+        // If this expression has any adjustments applied after the place expression,
+        // they may constitute reads.
+        if !self.typeck_results.borrow().expr_adjustments(expr).is_empty() {
```

Comment:
is this incorrect for `Deref` adjustments? 🤔

Reply:
A deref adjustment of a place expr would constitute a read, no? We'd then be derefing the place which requires a value.

Reply:
Although It's kinda moot -- we wouldn't ever be applying a deref adjustment to a place of type `!` anyways. The only adjustment we apply to `!` is `NeverToAny`. I could adjust this to check for just that I guess.

Reply:
but isn't this about the final type of an expr? i.e. in the `ty.is_never() && self.expr_constitutes_read(expr)` check in `check_expr_with_expectation_and_args`.

if you have some expr `foo()` of type `&!` which is implicitly deref'd to `!` (which probably can't happen), then we check whether the type *after adjustments* is `!` and whether that is the type of a place or value? I.e. it doesn't matter whether  `foo()` itself got converted to a value, it's whether the result after adjustments is a value 🤔

Reply:
Yeah, in which case I feel like it's most conservative to just remove this check.